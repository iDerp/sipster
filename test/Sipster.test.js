const sipster = require('../');
const mock_server = require('sip');
const debug = require('debug')('SIPSTER');
const EventEmitter = require('events');

const ENV = process.env;
const Domain = 'asterisk';
const MyID = '6001';
const CalleeID = '6002';
const HostAddress = ENV.ASTERISK_ADDRESS || '127.0.0.1';
const ServerPort = 5062;
const MyURI = `sip:${MyID}@${HostAddress}`;
const RegistrarURI = `sip:${HostAddress}:${ServerPort}`;
const CalleeURI = `sip:${CalleeID}@${HostAddress}:${ServerPort}`;
const ServerConfig = {
    port: ServerPort,
    logger: {
      send: (message, address) => debug(`SEND: ${JSON.stringify(message)}`),
      recv: (message, address) => debug(`RECV: ${JSON.stringify(message)}`)
    }
};
const accountConfig = {
    idUri: MyURI,
    regConfig: {
        registrarUri: RegistrarURI,
    },
    sipConfig: {
        authCreds: [{
            scheme: 'digest',
            realm: Domain,
            username: MyID,
            dataType: 0,
            data: MyID
        }],
    }
};

const TEST_HEADERS = [{
    hName: "x-test-header-name",
    hValue: "test-header-value"
}];
const TEST_HEADERS2 = [{
    hName: "x-test-header-name1",
    hValue: "test-header-value1"
}, {
    hName: "x-test-header-name2",
    hValue: "test-header-value2"
}];
const DEFAULT_REG_CONFIG = {
    registerOnAdd: true,
    contactParams: "",
    timeoutSec: 300,
    retryIntervalSec: 0,
    firstRetryIntervalSec: 0,
    randomRetryIntervalSec: 10,
    delayBeforeRefreshSec: 5,
    dropCallsOnFail: false,
    unregWaitMsec: 4000,
    proxyUse: 3,
};

const DEFAULT_SIP_CONFIG = {
    contactForced: "",
    contactParams: "",
    contactUriParams: "",
    authInitialEmpty: false,
    authInitialAlgorithm: "",
};
const sip_events = new EventEmitter();
const sdp = 'v=0\r\n'+
            'o=- 3736083706 3736083706 IN IP4 127.0.0.1\r\n'+
            's=-\r\n'+
            't=0 0\r\n'+
            'm=audio 4002 RTP/AVP 0\r\n'+
            'c=IN IP4 127.0.0.1\r\n'+
            'a=rtcp:4003 IN IP4 127.0.0.1\r\n'+
            'a=sendrecv\r\n'+
            'a=rtpmap:0 PCMU/8000\r\n';
const events = {
    registering: 0,
    unregistering: 0,
    registered: 0,
    unregistered: 0,
    state: 0,
    call: 0
};
const dialogs = {};
let account = {};
let call = {};
let cseq = 0;
let tag = 0;
let callId = 0;

/**
 * Complement the specified account config with default value.
 */
function makeAccountConfig(config) {
    if (config.regConfig) {
        const DEFAULT = DEFAULT_REG_CONFIG;
        config.regConfig = Object.assign(DEFAULT, config.regConfig);
    }
    if (config.sipConfig) {
        const DEFAULT = DEFAULT_SIP_CONFIG;
        config.sipConfig = Object.assign(DEFAULT, config.sipConfig);
    }
    return config;
}

function getTag() {
    return ++tag;
}

function getCallId() {
    return ++callId;
}

function getCseq() {
    return ++cseq;
}

function trim2check(message) {
    const m = JSON.parse(JSON.stringify(message));
    m.headers.via.forEach(via => {
        via.host = 'DNC';
        via.params.branch = 'DNC';
    });
    m.headers['call-id'] = 'DNC';
    m.headers.to.params.tag = 'DNC';
    m.headers.cseq.seq = 'DNC';
    m.headers.from.params.tag = 'DNC';
    m.headers['content-length'] = 'DNC';
    m.content = m.content.replace(/^o=-.+$/m, 'o=- DNC');
    m.content = m.content.replace(/\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/mg, 'DNC');
    return m;
}

beforeAll(() => {
    mock_server.start(ServerConfig, (request, remote) => {
        debug('request: ', mock_server.stringify(request));
        sip_events.emit(request.method, request, remote);
    });
});

afterAll(() => {
    mock_server.stop();
});

describe('sipster', () => {
    test ('check static properties', () => {
        expect(sipster.version).toMatchSnapshot();
    });
    test ('initialize pjsip', () => {
        const configs = { logConfig: { level: 3, consoleLevel: 3}};
        sipster.init(configs);
        expect(sipster.config).toMatchSnapshot();
        expect(sipster.mediaActivePorts).toMatchSnapshot();
        expect(sipster.mediaMaxPorts).toMatchSnapshot();
        expect(sipster.state).toMatchSnapshot();
    });
    test ('set up a transport to listen for incoming connections, defaults to UDP', () => {
        transport = new sipster.Transport({ port: 5060 });
        const info = transport.getInfo();
        info.info = undefined;
        info.localName = undefined;
        expect(info).toMatchSnapshot();
        expect(transport.enabled).toMatchSnapshot();
    });
    test ('set up a SIP account, we need at least one', () => {
        // this sets up an account for calls coming from 192.168.100.10
        account = new sipster.Account(makeAccountConfig(accountConfig));
        account.on('registering', () => events.registering++);
        account.on('unregistering', () => events.unregistering++);
        account.on('registered', () => events.registered++);
        account.on('unregistered', () => events.unregistered++);
        account.on('state', (active, statusCode) => events.state++);
        account.on('call', (info, call) => events.call++);
        expect(account.getInfo()).toMatchSnapshot();
        expect(account.valid).toMatchSnapshot();
        expect(account.default).toMatchSnapshot();
    });
    test ('make an audio player', () => {
        player = sipster.createPlayer(`${__dirname}/waves/sound.wav`);
    });
    test ('make an audio recorder', () => {
        recorder = sipster.createRecorder(`${__dirname}/waves/call.wav`);
    });
    test ('finalize the pjsip initialization phase', done => {
        sip_events.once('REGISTER', request => {
            const response = mock_server.makeResponse(request, 200, 'OK');
            mock_server.send(response);
        });
        account.once('registered', () => {
            debug('an account has been registered');
            return done();
        });
        sipster.start();
    });
//
// Outbound call
//
    test ('make an outbound call', done => {
        debug('making an outbound call');
        sip_events.once('INVITE', request => {
            expect(trim2check(request)).toMatchSnapshot();
            const response = mock_server.makeResponse(request, 200, 'OK');
            response.headers['content-type'] = 'application/sdp';
            response.content = sdp;
            mock_server.send(response);
        });
        call = account.makeCall(CalleeURI, TEST_HEADERS);
        call.on('state', state => {
            if (state === 'confirmed') {
                debug('the calling confirmed');
                call.removeAllListeners('state');
                return done();
            }
        });
    });
    test ('hang up the outbound call', done => {
        debug('hanging up the outbound call');
        sip_events.once('BYE', request => {
            const response = mock_server.makeResponse(request, 200, 'OK');
            mock_server.send(response);
        });
        call.on('state', state => {
            if (state === 'disconnected') {
                debug('the calling has been disconnected');
                call.removeAllListeners('state');
                call = null;
                return done();
            }
        });
        call.hangup();
    });

//
// Outbound call #2 w/o custom headers
//
    test ('make an outbound call #2', done => {
        debug('making an outbound call #2');
        sip_events.once('INVITE', request => {
            expect(trim2check(request)).toMatchSnapshot();
            const response = mock_server.makeResponse(request, 200, 'OK');
            response.headers['content-type'] = 'application/sdp';
            response.content = sdp;
            mock_server.send(response);
        });
        call = account.makeCall(CalleeURI);
        call.on('state', state => {
            if (state === 'confirmed') {
                debug('the calling confirmed');
                call.removeAllListeners('state');
                return done();
            }
        });
    });
    test ('hang up the outbound call #2', done => {
        debug('hanging up the outbound call #2');
        sip_events.once('BYE', request => {
            const response = mock_server.makeResponse(request, 200, 'OK');
            mock_server.send(response);
        });
        call.on('state', state => {
            if (state === 'disconnected') {
                debug('the calling has been disconnected');
                call.removeAllListeners('state');
                call = null;
                return done();
            }
        });
        call.hangup();
    });

//
// Inbound call with custom headers
//
    test ('receive an inbound call', done => {
        debug('receive an inboud call');
        account.on('call', (info, inbound_call) => {
            debug('call', inbound_call);
            inbound_call.answer(TEST_HEADERS2);
            call = inbound_call;
        });
        mock_server.send({
            method: 'INVITE',
            uri: MyURI,
            headers: {
                to: {uri: MyURI},
                from: {uri: CalleeURI, params: {tag: getTag()}},
                'call-id': getCallId(),
                cseq: {method: 'INVITE', seq: getCseq()},
                'content-type': 'application/sdp',
                contact: [{uri: CalleeURI}]
            },
            content:
                'v=0\r\n'+
                'o=- 13374 13374 IN IP4 127.0.0.1\r\n'+
                's=-\r\n'+
                'c=IN IP4 127.0.0.1\r\n'+
                't=0 0\r\n'+
                'm=audio 4000 RTP/AVP 0\r\n'+
                'a=rtpmap:0 PCMU/8000\r\n'+
                'a=sendrecv\r\n'
        }, response => {
            expect(trim2check(response)).toMatchSnapshot();
            debug('an inbound call has been established');
            mock_server.send({
                method: 'ACK',
                uri: MyURI,
                headers: {
                    from: response.headers.from,
                    to: response.headers.to,
                    cseq: {method: 'ACK', seq: response.headers.cseq.seq},
                    'call-id': response.headers['call-id'],
                    via: [response.headers.via[0]],
                    'max-forwards': 70
                }
            });
            done();
        });
    });
    test ('hang up the inbound call', done => {
        debug('hanging up the inbound call');
        sip_events.once('BYE', request => {
            const response = mock_server.makeResponse(request, 200, 'OK');
            mock_server.send(response);
        });
        call.on('state', state => {
            if (state === 'disconnected') {
                debug('the calling has been disconnected');
                call.removeAllListeners('state');
                return done();
            }
        });
        call.hangup();
    });
//
// Unregister
//
    test ('unregister account', done => {
        debug('unregister the account');
        sip_events.once('REGISTER', request => {
            const response = mock_server.makeResponse(request, 200, 'OK');
            mock_server.send(response);
        });
        account.setRegistration(false);
        account.once('unregistered', () => {
            debug('the account has been unregistered');
            expect(sipster.state).toMatchSnapshot();
            expect(events).toMatchSnapshot();
            return done();
        });
    });
});
