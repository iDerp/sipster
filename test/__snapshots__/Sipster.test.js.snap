// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sipster check static properties 1`] = `
Object {
  "full": "2.7.2",
  "major": 2,
  "minor": 7,
  "numeric": 34013696,
  "rev": 2,
  "suffix": "",
}
`;

exports[`sipster hang up the inbound call 1`] = `
Object {
  "content": "v=0
o=- DNC
s=pjmedia
b=AS:84
t=0 0
a=X-nat:0
m=audio 4004 RTP/AVP 0
c=IN IP4 DNC
b=TIAS:64000
a=rtcp:4005 IN IP4 DNC
a=sendrecv
a=rtpmap:0 PCMU/8000
",
  "headers": Object {
    "allow": "PRACK, INVITE, ACK, BYE, CANCEL, UPDATE, INFO, SUBSCRIBE, NOTIFY, REFER, MESSAGE, OPTIONS",
    "call-id": "DNC",
    "contact": Array [
      Object {
        "params": Object {},
        "uri": "sip:6001@127.0.0.1:5060;ob",
      },
    ],
    "content-length": "DNC",
    "content-type": "application/sdp",
    "cseq": Object {
      "method": "INVITE",
      "seq": "DNC",
    },
    "from": Object {
      "params": Object {
        "tag": "DNC",
      },
      "uri": "sip:6002@127.0.0.1",
    },
    "supported": "replaces, 100rel, timer, norefersub",
    "to": Object {
      "params": Object {
        "tag": "DNC",
      },
      "uri": "sip:6001@127.0.0.1",
    },
    "via": Array [
      Object {
        "host": "DNC",
        "params": Object {
          "branch": "DNC",
          "received": "127.0.0.1",
          "rport": "5062",
        },
        "port": 5062,
        "protocol": "UDP",
        "version": "2.0",
      },
    ],
    "x-test-header-name1": "test-header-value1",
    "x-test-header-name2": "test-header-value2",
  },
  "reason": "OK",
  "status": 200,
  "version": "2.0",
}
`;

exports[`sipster initialize pjsip 1`] = `
Object {
  "EpConfig": Object {
    "LogConfig": Object {
      "consoleLevel": 3,
      "decor": 25328,
      "fileFlags": 0,
      "filename": "",
      "level": 3,
      "msgLogging": 1,
    },
    "MediaConfig": Object {
      "audioFramePtime": 20,
      "channelCount": 1,
      "clockRate": 16000,
      "ecOptions": 0,
      "ecTailLen": 200,
      "hasIoqueue": true,
      "ilbcMode": 30,
      "jbInit": -1,
      "jbMax": -1,
      "jbMaxPre": -1,
      "jbMinPre": -1,
      "maxMediaPorts": 254,
      "noVad": false,
      "ptime": 0,
      "quality": 8,
      "rxDropPct": 0,
      "sndAutoCloseTime": 1,
      "sndClockRate": 0,
      "sndPlayLatency": 140,
      "sndRecLatency": 100,
      "threadCnt": 1,
      "txDropPct": 0,
      "vidPreviewEnableNative": true,
    },
    "UaConfig": Object {
      "mainThreadOnly": false,
      "maxCalls": 4,
      "mwiUnsolicitedEnabled": true,
      "nameserver": Array [],
      "natTypeInSdp": 1,
      "stunIgnoreFailure": true,
      "stunServer": Array [],
      "stunTryIpv6": false,
      "threadCnt": 1,
      "userAgent": "",
    },
  },
}
`;

exports[`sipster initialize pjsip 2`] = `1`;

exports[`sipster initialize pjsip 3`] = `254`;

exports[`sipster initialize pjsip 4`] = `"init"`;

exports[`sipster make an outbound call #2 1`] = `
Object {
  "content": "v=0
o=- DNC
s=pjmedia
b=AS:84
t=0 0
a=X-nat:0
m=audio 4002 RTP/AVP 0 8 96
c=IN IP4 DNC
b=TIAS:64000
a=rtcp:4003 IN IP4 DNC
a=sendrecv
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:96 telephone-event/8000
a=fmtp:96 0-16
",
  "headers": Object {
    "allow": "PRACK, INVITE, ACK, BYE, CANCEL, UPDATE, INFO, SUBSCRIBE, NOTIFY, REFER, MESSAGE, OPTIONS",
    "call-id": "DNC",
    "contact": Array [
      Object {
        "params": Object {},
        "uri": "sip:6001@127.0.0.1:5060;ob",
      },
    ],
    "content-length": "DNC",
    "content-type": "application/sdp",
    "cseq": Object {
      "method": "INVITE",
      "seq": "DNC",
    },
    "from": Object {
      "params": Object {
        "tag": "DNC",
      },
      "uri": "sip:6001@127.0.0.1",
    },
    "max-forwards": "70",
    "min-se": "90",
    "session-expires": "1800",
    "supported": "replaces, 100rel, timer, norefersub",
    "to": Object {
      "params": Object {
        "tag": "DNC",
      },
      "uri": "sip:6002@127.0.0.1",
    },
    "via": Array [
      Object {
        "host": "DNC",
        "params": Object {
          "branch": "DNC",
          "received": "127.0.0.1",
          "rport": 5060,
        },
        "port": 5060,
        "protocol": "UDP",
        "version": "2.0",
      },
    ],
  },
  "method": "INVITE",
  "uri": "sip:6002@127.0.0.1:5062",
  "version": "2.0",
}
`;

exports[`sipster make an outbound call 1`] = `
Object {
  "content": "v=0
o=- DNC
s=pjmedia
b=AS:84
t=0 0
a=X-nat:0
m=audio 4000 RTP/AVP 0 8 96
c=IN IP4 DNC
b=TIAS:64000
a=rtcp:4001 IN IP4 DNC
a=sendrecv
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:96 telephone-event/8000
a=fmtp:96 0-16
",
  "headers": Object {
    "allow": "PRACK, INVITE, ACK, BYE, CANCEL, UPDATE, INFO, SUBSCRIBE, NOTIFY, REFER, MESSAGE, OPTIONS",
    "call-id": "DNC",
    "contact": Array [
      Object {
        "params": Object {},
        "uri": "sip:6001@127.0.0.1:5060;ob",
      },
    ],
    "content-length": "DNC",
    "content-type": "application/sdp",
    "cseq": Object {
      "method": "INVITE",
      "seq": "DNC",
    },
    "from": Object {
      "params": Object {
        "tag": "DNC",
      },
      "uri": "sip:6001@127.0.0.1",
    },
    "max-forwards": "70",
    "min-se": "90",
    "session-expires": "1800",
    "supported": "replaces, 100rel, timer, norefersub",
    "to": Object {
      "params": Object {
        "tag": "DNC",
      },
      "uri": "sip:6002@127.0.0.1",
    },
    "via": Array [
      Object {
        "host": "DNC",
        "params": Object {
          "branch": "DNC",
          "received": "127.0.0.1",
          "rport": 5060,
        },
        "port": 5060,
        "protocol": "UDP",
        "version": "2.0",
      },
    ],
    "x-test-header-name": "test-header-value",
  },
  "method": "INVITE",
  "uri": "sip:6002@127.0.0.1:5062",
  "version": "2.0",
}
`;

exports[`sipster receive an inbound call 1`] = `
Object {
  "content": "",
  "headers": Object {
    "call-id": "DNC",
    "content-length": "DNC",
    "cseq": Object {
      "method": "INVITE",
      "seq": "DNC",
    },
    "from": Object {
      "params": Object {
        "tag": "DNC",
      },
      "uri": "sip:6002@127.0.0.1",
    },
    "to": Object {
      "params": Object {
        "tag": "DNC",
      },
      "uri": "sip:6001@127.0.0.1",
    },
    "via": Array [
      Object {
        "host": "DNC",
        "params": Object {
          "branch": "DNC",
          "received": "127.0.0.1",
          "rport": "5062",
        },
        "port": 5062,
        "protocol": "UDP",
        "version": "2.0",
      },
    ],
  },
  "reason": "Trying",
  "status": 100,
  "version": "2.0",
}
`;

exports[`sipster set up a SIP account, we need at least one 1`] = `
Object {
  "regExpiresSec": 0,
  "regIsActive": false,
  "regIsConfigured": true,
  "uri": "sip:6001@127.0.0.1",
}
`;

exports[`sipster set up a SIP account, we need at least one 2`] = `true`;

exports[`sipster set up a SIP account, we need at least one 3`] = `true`;

exports[`sipster set up a transport to listen for incoming connections, defaults to UDP 1`] = `
Object {
  "flags": 4,
  "info": undefined,
  "localAddress": "0.0.0.0:5060",
  "localName": undefined,
  "type": "UDP",
  "usageCount": 1,
}
`;

exports[`sipster set up a transport to listen for incoming connections, defaults to UDP 2`] = `true`;

exports[`sipster unregister account 1`] = `"running"`;

exports[`sipster unregister account 2`] = `
Object {
  "call": 1,
  "registered": 1,
  "registering": 1,
  "state": 1,
  "unregistered": 1,
  "unregistering": 1,
}
`;
